### PR TITLE
sd-network: DHCPv4 - Add user class, extra option and send generic o…

### PR DESCRIPTION
**Pull request by ssahani**
…ptions in request state

closes https://github.com/systemd/systemd/issues/15470

### CODE REVIEW

**Comment by ssahani** [comment]
Cc @yuwata

File: src/libsystemd-network/sd-dhcp-client.c [hunk 1]
```diff
@@ -1064,6 +1066,30 @@ static int client_send_request(sd_dhcp_client *client) {
...
+        }
+
+        if (!ordered_hashmap_isempty(client->vendor_options)) {
+                r = dhcp_option_append(
+                                &request->dhcp, optlen, &optoffset, 0,
+                                SD_DHCP_OPTION_VENDOR_SPECIFIC,
+                                ordered_hashmap_size(client->vendor_options), client->vendor_options);
+                if (r < 0)
+                        return r;
+        }
```
Comment (poettering) [inline src/libsystemd-network/sd-dhcp-client.c]: hmm, the very same lines already exist in client_send_discover(). Maybe add some common shared helper function client_append_common_discover_request_options()? i.e. a helper that appens all options included in both request + discover?

Comment (ssahani) [reply src/libsystemd-network/sd-dhcp-client.c]: Ok .

**Comment by ssahani** [comment]
Updated. Thanks for the review.

**Comment by ssahani** [comment]
Updated thanks for the review . PTAL.

File: src/libsystemd-network/sd-dhcp-client.c [hunk 2]
```diff
@@ -928,26 +904,61 @@ static int client_send_discover(sd_dhcp_client *client) {
...
+
+        r = client_append_common_discover_request_options(client, discover, &optoffset, optlen);
         if (r < 0)
                 return r;
 
+        r = dhcp_option_append(&discover->dhcp, optlen, &optoffset, 0,
+                               SD_DHCP_OPTION_END, 0, NULL);
+
```
Comment (poettering) [inline src/libsystemd-network/sd-dhcp-client.c]: missing check for failure

**Comment by ssahani** [comment]
FIxed and pushed. Thanks !

**Pull request merged**